# Data Analysis

## Setup

In [1]:
import pandas as pd
import numpy as np

path_preproccesed = '../data/preprocessed.parquet'

In [2]:
df = pd.read_parquet(path_preproccesed, engine='fastparquet')
df.head()

property_id  building  date_sale  type  property#     area      price  \
0         1030         1 2005-11-01     1         30   743.09  246172.68   
1         1029         1 2005-10-01     1         29   756.21  246331.90   
2         2002         2 2007-07-01     1          2   587.28  209280.91   
3         2031         2 2007-12-01     1         31  1604.75  452667.01   
4         1049         1 2004-11-01     1         49  1375.45  467083.31   

   status customerid  entity  ... birth_date  sex country       state purpose  \
0       1      C0028     0.0  ... 1986-06-21  1.0     USA  California     0.0   
1       1      C0027     0.0  ... 1983-02-24  1.0     USA  California     0.0   
2       1      C0112     0.0  ... 1985-12-27  0.0     USA  California     0.0   
3       1      C0160     0.0  ... 1985-12-27  0.0     USA  California     1.0   
4       1      C0014     0.0  ... 1979-05-15  1.0     USA  California     0.0   

  deal_satisfaction  mortgage  source_Agency  source_Client  source_Website  
0               5.0       0.0          False          False            True  
1               5.0       0.0          False          False            True  
2               1.0       1.0          False           True           False  
3               3.0       1.0          False          False            True  
4               4.0       0.0           True          False           False  

[5 rows x 22 columns]

In [3]:
df.shape

(267, 22)

In [4]:
df.dtypes

property_id                   int64
building                      int64
date_sale            datetime64[ns]
type                          int64
property#                     int64
area                        float64
price                       float64
status                        int64
customerid                   object
entity                      float64
name                         object
surname                      object
birth_date           datetime64[ns]
sex                         float64
country                      object
state                        object
purpose                     float64
deal_satisfaction           float64
mortgage                    float64
source_Agency                  bool
source_Client                  bool
source_Website                 bool
dtype: object

---

## Analysis

### **Descriptive Statistics**

The aim is to understand the distribution of the key variables to answer the following questions:
- What are the sales and overall performance by building type?
- What are the sales and overall performance by country and state?

### `Building`

There are 5 types of buildings in the dataset. We will to go step by step.

In [5]:
set(df['building'])

{1, 2, 3, 4, 5}

In order to examine the totals by building types, we study the total number of sold properties and how many of them had mortgages per building type.

In [6]:
buildings_df = df[['building', 'mortgage', 'status']]

In [7]:
sold_buildings_df = buildings_df.loc[df['status'] == 1].drop('status', axis=1).copy()
sold_buildings_df

building  mortgage
0           1       0.0
1           1       0.0
2           2       1.0
3           2       1.0
4           1       0.0
..        ...       ...
191         1       0.0
192         1       0.0
193         1       0.0
194         1       0.0
195         1       0.0

[195 rows x 2 columns]

In [8]:
sold_buildings_df.groupby('building').count()

mortgage
building          
1               46
2               54
3               53
4               23
5               19

This indicates that there are 54 type 2 buildings sold.

In [9]:
sold_buildings_df.groupby('building').sum()

mortgage
building          
1             14.0
2             18.0
3             15.0
4              9.0
5              6.0

It is important to remember that each `Yes` in the column `mortgage` is 1. Consequently, we can observe that type 2 buildings have the most amount of mortgages, whereas type 5 building the least amount.

Following this, we are going to determine the average values of `area`, `price`, and `deal_satisfaction` per building type.

In [10]:
buildings_df = df[['building', 'area', 'price', 'deal_satisfaction', 'status']]

sold_buildings_df = buildings_df.loc[df['status'] == 1].drop('status', axis=1).copy()
sold_buildings_df

building     area      price  deal_satisfaction
0           1   743.09  246172.68                5.0
1           1   756.21  246331.90                5.0
2           2   587.28  209280.91                1.0
3           2  1604.75  452667.01                3.0
4           1  1375.45  467083.31                4.0
..        ...      ...        ...                ...
191         1   743.09  210745.17                5.0
192         1  1160.36  331154.88                5.0
193         1   625.80  204434.68                5.0
194         1   756.21  189194.31                5.0
195         1   625.80  204027.09                5.0

[195 rows x 4 columns]

In [11]:
sold_buildings_df.groupby('building').mean()

area          price  deal_satisfaction
building                                              
1         927.085217  275218.989348           3.630435
2         922.192037  278966.224630           3.518519
3         870.257170  259241.115660           3.566038
4         933.577826  274722.053913           3.869565
5         812.309474  250373.861579           3.526316

Note that the highest `deal_satisfaction` corresponds to building type 4, which has the greatest mean `area` and isn't the most expensive (according to mean `price`).

### `Country`

We are going to repeat the same analysis we did for the building types. 

In [12]:
sold_df = df[df['status'] == 1]
sold_df

property_id  building  date_sale  type  property#     area      price  \
0           1030         1 2005-11-01     1         30   743.09  246172.68   
1           1029         1 2005-10-01     1         29   756.21  246331.90   
2           2002         2 2007-07-01     1          2   587.28  209280.91   
3           2031         2 2007-12-01     1         31  1604.75  452667.01   
4           1049         1 2004-11-01     1         49  1375.45  467083.31   
..           ...       ...        ...   ...        ...      ...        ...   
191         1022         1 2005-07-01     1         22   743.09  210745.17   
192         1028         1 2005-07-01     1         28  1160.36  331154.88   
193         1034         1 2005-07-01     1         34   625.80  204434.68   
194         1037         1 2005-07-01     1         37   756.21  189194.31   
195         1042         1 2005-07-01     1         42   625.80  204027.09   

     status customerid  entity  ... birth_date  sex country       state  \
0         1      C0028     0.0  ... 1986-06-21  1.0     USA  California   
1         1      C0027     0.0  ... 1983-02-24  1.0     USA  California   
2         1      C0112     0.0  ... 1985-12-27  0.0     USA  California   
3         1      C0160     0.0  ... 1985-12-27  0.0     USA  California   
4         1      C0014     0.0  ... 1979-05-15  1.0     USA  California   
..      ...        ...     ...  ...        ...  ...     ...         ...   
191       1      C0174     1.0  ...        NaT  NaN     USA  California   
192       1      C0174     1.0  ...        NaT  NaN     USA  California   
193       1      C0174     1.0  ...        NaT  NaN     USA  California   
194       1      C0174     1.0  ...        NaT  NaN     USA  California   
195       1      C0174     1.0  ...        NaT  NaN     USA  California   

    purpose deal_satisfaction  mortgage  source_Agency  source_Client  \
0       0.0               5.0       0.0          False          False   
1       0.0               5.0       0.0          False          False   
2       0.0               1.0       1.0          False           True   
3       1.0               3.0       1.0          False          False   
4       0.0               4.0       0.0           True          False   
..      ...               ...       ...            ...            ...   
191     1.0               5.0       0.0          False          False   
192     1.0               5.0       0.0          False          False   
193     1.0               5.0       0.0          False          False   
194     1.0               5.0       0.0          False          False   
195     1.0               5.0       0.0          False          False   

     source_Website  
0              True  
1              True  
2             False  
3              True  
4             False  
..              ...  
191            True  
192            True  
193            True  
194            True  
195            True  

[195 rows x 22 columns]

In [13]:
important_columns = ['country','area', 'price', 'deal_satisfaction','mortgage']

country_sold_df = sold_df[important_columns]

In [14]:
temp_view = country_sold_df.drop(['area', 'price', 'deal_satisfaction'], axis=1)
temp_view.groupby('country').sum()

mortgage
country          
Belgium       0.0
Canada        0.0
Denmark       0.0
Germany       0.0
Mexico        0.0
Russia        1.0
UK            0.0
USA          61.0

From the results, we can see the US has the most amount of mortgages.

In [15]:
temp_view = country_sold_df.drop(['mortgage'], axis=1)
temp_view.groupby('country').mean()

area          price  deal_satisfaction
country                                               
Belgium   852.730000  229075.470000           3.000000
Canada    917.382857  274069.384286           5.000000
Denmark   785.480000  257183.480000           1.000000
Germany   743.410000  205098.210000           5.000000
Mexico   1283.450000  338181.180000           3.000000
Russia    903.757500  278828.835000           2.500000
UK        739.480000  220142.680000           4.000000
USA       900.794463  270096.266554           3.581921

Keep in mind that we don't possess enough data for many countries like Mexico or Germany.

### `State`

In [16]:
sold_us_df = df[(df['status'] == 1) & (df['country'] == 'USA')]

important_columns = ['state','area', 'price', 'deal_satisfaction','mortgage']
state_sold_df = sold_us_df[important_columns]

In [17]:
temp_view = state_sold_df.drop(['area', 'price', 'deal_satisfaction'], axis=1)
temp_view.groupby('state').sum()

mortgage
state               
Arizona          3.0
California      41.0
Colorado         5.0
Kansas           0.0
Nevada           8.0
Oregon           1.0
Utah             1.0
Virginia         2.0
Wyoming          0.0

In [18]:
temp_view = state_sold_df.drop(['mortgage'], axis=1)
temp_view.groupby('state').mean()

area          price  deal_satisfaction
state                                                   
Arizona     848.533333  265079.216667           3.222222
California  936.684492  279852.950763           3.737288
Colorado    777.010909  233662.461818           3.181818
Kansas      923.210000  252185.990000           3.000000
Nevada      921.692941  277649.329412           3.235294
Oregon      766.954545  236593.322727           3.090909
Utah        814.986000  237755.280000           3.200000
Virginia    730.712500  215144.187500           4.500000
Wyoming     701.660000  204286.670000           3.000000

As shown in the input, California has the most amount of mortgages and the deal satisfaction is 3.737. The best deal satisfaction is found in Virginia. It's worth noting that the dataset is unbalanced and for some states we don't have enough data.

### **In-depth Analysis**

The objective of this analysis is to uncover trends, correlations, and hidden insights, to find out more about the relationship between `age` and `price`.

In [19]:
analysis_df = df.copy()
analysis_df.head()

property_id  building  date_sale  type  property#     area      price  \
0         1030         1 2005-11-01     1         30   743.09  246172.68   
1         1029         1 2005-10-01     1         29   756.21  246331.90   
2         2002         2 2007-07-01     1          2   587.28  209280.91   
3         2031         2 2007-12-01     1         31  1604.75  452667.01   
4         1049         1 2004-11-01     1         49  1375.45  467083.31   

   status customerid  entity  ... birth_date  sex country       state purpose  \
0       1      C0028     0.0  ... 1986-06-21  1.0     USA  California     0.0   
1       1      C0027     0.0  ... 1983-02-24  1.0     USA  California     0.0   
2       1      C0112     0.0  ... 1985-12-27  0.0     USA  California     0.0   
3       1      C0160     0.0  ... 1985-12-27  0.0     USA  California     1.0   
4       1      C0014     0.0  ... 1979-05-15  1.0     USA  California     0.0   

  deal_satisfaction  mortgage  source_Agency  source_Client  source_Website  
0               5.0       0.0          False          False            True  
1               5.0       0.0          False          False            True  
2               1.0       1.0          False           True           False  
3               3.0       1.0          False          False            True  
4               4.0       0.0           True          False           False  

[5 rows x 22 columns]

### `age`

We begin by determining the age of customers at the time of sale.

In [20]:
analysis_df['age'] = (2025 - analysis_df['birth_date'].dt.year).astype('Int64')
analysis_df['age']

0        39
1        42
2        40
3        40
4        46
       ... 
262    <NA>
263    <NA>
264    <NA>
265    <NA>
266    <NA>
Name: age, Length: 267, dtype: Int64

Having completed this step, we proceed to create age intervals.

In [21]:
bins = [19, 25, 31, 36, 42, 48, 54, 59, 65, 71, 76]
labels = [f'({bins[i]}, {bins[i+1]}]' for i in range(len(bins)-1)]

analysis_df['age_interval'] = pd.cut(
    analysis_df['age'],
    bins=bins,
    labels=labels,
    right=True,    
    include_lowest=False
)

Moving on, we determine how many properties have been sold by age intervals

In [22]:
sold_analysis_df = analysis_df[analysis_df['status'] == 1].copy()

sold_analysis_df[['age_interval', 'status']].groupby('age_interval').count()

/tmp/ipykernel_19021/2084252935.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  sold_analysis_df[['age_interval', 'status']].groupby('age_interval').count()


status
age_interval        
(19, 25]           0
(25, 31]           0
(31, 36]           0
(36, 42]           4
(42, 48]          14
(48, 54]          26
(54, 59]          27
(59, 65]          26
(65, 71]          27
(71, 76]          15

This suggest that young adults aren't able to afford a property and most customers are middle age adults or senior adults.

Let's now focus on studying the relationship between `age` and `price`.

In [23]:
covariance = sold_analysis_df['price'].cov(sold_analysis_df['age'])
correlation = sold_analysis_df['price'].corr(sold_analysis_df['age'])

print(f'Covariance = {covariance:0.2f}')
print(f'Correlation = {correlation:0.2f}')

Covariance = -169748.66
Correlation = -0.17


The negative covariance suggests that the two variables tend to move in opposite directions. However, the correlation value of -0.17 indicates that the linear relationship between them is very weak, almost nonexistent.